[2단계 - 자동차 경주] 황의찬 미션 제출합니다.

pine14 opened the pull request:
2단계 미션 제출합니다.

Car 객체가 너무 많은 책임을 가지고 있는 것 같아서, 객체지향생활체조원칙 에 따라 원시값과 문자열을 포장했는데, 잘 한것인지 모르겠습니다..😂

추가적으로 random 값 테스트 하는것에 대한 방법을 모르겠습니다.
구글에 찾아보니 인터페이스를 이용해 테스트를 진행한 글들을 많이 봐왔는데, 그렇게 적용하면 Car 클래스의 인스턴스가 3개가 만들어져 객체지향생활체조원칙 규칙 7번에도 어긋나서 일단은 랜덤 숫자를 임의로 지정했을 때 자동차가 정지하는지, 이동하는지의 테스트만 진행했습니다.

열심히 리뷰해 주셔서 감사드립니다!

Commit:
refactor : 표준 예외를 사용하기

Exception, RuntimeException, Throwable, Error는 직접 재사용하지 말자.
이 예외들은 다른 예외들의 상위 클래스이므로, 즉 여러 성격의 예외들을 포함하는 클래스이므로 안정적으로
테스트 할 수 없다. - 출처 : 이펙티브 자바
의 글을 보고 RuntimeException을 IllegalArgumentException으로 리팩터링했습니다.

Commit:
docs : 문자열 계산기 요구사항 수정

Commit:
refactor : 상수이므로 대문자 표기

상수의 경우 대문자로 표기하는 컨벤션이 있으므로 변경했습니다.

Commit:
MethodSource를 사용하는 코드 CsvSource 사용하기

CsvSoucre를 사용해서 (4, true), (3,false) 와 같이
4일 때 전진, 3일 때 멈춤 을 테스트할 수 있습니다. 따라서 randomNumberGenerate 메서드를 제거할 수 있습니다.

Commit:
refactor : 상수 접근 제어 변경

테스트 코드가 프로덕션 코드에 영향을 주어서는 안됩니다.
또한 TDD를 제대로 수행했다면 테스트를 먼저 작성했을 테니, 테스트 클래스에서 사용할 상수를 먼저
선언했을 것이므로 접근 제어를 public -> private으로 변경했습니다.

Commit:
refactor : 클래스 내부 첫 줄 공백 제거

Clean Code 5장을 참고해서 클래스 내부 첫 줄의 공백을 제거하는 방식으로 통일시켰습니다.

Commit:
refactor : 파일의 마지막에 새 줄 추가

캠퍼스 핵데이 Java 코딩 컨벤션에 따라 파일의 마지막에 새 줄을 추가했습니다.

Commit:
refactor : 클래스를 정의하는 표준 자바 관례에 따라 가독성 높이기

정적 공개(public) 상수, 정적 비공개 변수, 비공개 인스턴스 변수 순으로 나오도록 리팩토링했습니다.

Commit:
refactor : 클래스 내에서 메서드를 위치하는 방식 변경

Clean Code 10장과 Standard Java Convention을 보고
비공개 함수는 자신이 호출하는 공개 함수 직후에 넣는 방식을 사용했습니다.
가독성을 더 중요하게 생각했습니다.

Commit:
refactor : 클래스 이름 변경

view의 역할을 하는 클래스 이름에 맞게 변경했습니다.

Commit:
refactor : 메서드 순서 변경해서 가독성 높이기

비공개 함수를 자신이 호출하는 공개 함수 직후에 위치시켰습니다.

Commit:
refactor : isWinner 메서드 수정

if문을 없애고 바로 return하는 방식으로 리팩토링했습니다.

Commit:
refactor : repeat 함수를 이용해 반복 처리

Java 11에서 도입된 repeat 함수를 도입해 반복 코드를 리팩토링했습니다.

Commit:
핵심 비즈니스 로직을 가지는 객체 domain 패키지로 변경

리팩터링 요구사항에 따라 핵심 비지니스 로직을 가지는 객체를 domain 패키지로 변경했습니다.

Commit:
원시 타입 값 포장

기존의 Car 객체는 이름에 대한 유효성 검증에 대한 책임까지 가지고 있으므로
각각 Name, Position 클래스로 포장했습니다.
원시 타입 값을 포장하게 되면, 그 변수가 의미하는 바를 명확히 나타낼 수 있고 유지보수에도 많은 도움이 됩니다.

Commit:
refactor : 검증 방식 변경

생성자에서 검증을 진행합니다.
이름의 경우 중복 처리는 일급 컬렉션에서, 나머지는 분리한 Name 클래스에서 진행합니다.

Commit:
refactor : 자동차 생성 테스트 수정

원시값과 문자열을 포장했기 때문에 테스트 클래스를 맞게 수정했습니다.

Commit:
test : 이름 검증 테스트 작성

Name 클래스에서 이름 검증을 담당하므로 그에 해당하는 테스트를 진행했습니다.

Commit:
refactor : Displaynamegeneration 삭제

underscore 함수가 없으므로 불필요한 DisplaynameGeneration 애노테이션을 제거했습니다.

Commit:
test : 자동차의 위치 테스트

자동차의 위치는 음수일 수 없고, 자동차가 전진하면 위치가 1 증가해야 합니다.

Commit:
refactor : NameValidator, NameValidatorTest 삭제

Name 클래스의 생성자와 일급 컬렉션에서 검증을 진행하기 때문에 사용하지 않는 NameValidator 클래스는 삭제해주었습니다.

Commit:
refactor : CarsTest 수정

일급 컬렉션 테스트의 인스턴스 초기화를 수정했습니다.

Commit:
refactor : 랜던 값 생성 변경

skullkim commented on src/main/java/calculator/Calculator.java:
```diff
@@ -5,13 +5,12 @@
 	private static final int INPUT_GROUP = 2;
 	private static final int NEGATIVE_NUMBER_STANDARD = 0;
-	private static final Pattern pattern = Pattern.compile(CUSTOM_SPLIT_REGEX);
+	private static final Pattern COMPILE_CUSTOM_SPLIT_REGEX = Pattern.compile(CUSTOM_SPLIT_REGEX);
```
변수명은 명사 또는 명사구여야 합니다. 그런데 'compile'은 동사죠. Pattern.compile()로 이미 컴파일을 진행 한 상태를 저장하는 것이니 'COMPILED_CUSTOM_SPLIT_REGEX'는 어떨까요🧐?

pine14 replied:
클린 코드 2장을 읽어봤습니다! 
[클래스 이름과 객체 이름은 명사 또는 명사구를 사용하라] 라는 말이 있으므로 COMPILED_CUSTOM_SPLIT_REGEX 라는 
컴파일된 사용자 정의 분할 정규식 이름이 적합한 것 같습니다!

skullkim commented on src/main/java/racingcar/controller/RaceGameController.java:
```diff
@@ -1,63 +1,67 @@
 	private static void printResultHeader() {
-		OutputData.printBlankLine();
-		OutputData.printGameResultHeader();
+		OutputView.printBlankLine();
```
이 메서드만 static인 이유가 있을까요?

pine14 replied:
static일 필요가 없었습니다! 😂 제거하고 커밋하겠습니다.
다른 곳에서 `RaceGameController.printResultHeader();` 와 같이 호출하는 곳이 없네요!

https://chan9.tistory.com/168
리뷰어님의 리뷰를 인용했습니다! 👍

skullkim commented on src/main/java/racingcar/util/validator/NumberValidator.java:
```diff
@@ -1,7 +1,6 @@
 
 public class NumberValidator {
-
 	private static final String MIN_ATTEMPT_NUMBER_MESSAGE = "시도 횟수는 1회 이상이어야 합니다.";
```
시도 횟수를 정하는 것은 도메인 로직이라는 생각이 드네요. 그런데 현재 NumberValidator를 util의 NumberGenerator에서 사용하고 있고, NumberGenerator를 Controller에서 사용하고 있네요. 이로직을 도메인으로 옮길 수 있는 방법이 없을까요🧐?

pine14 replied:
AttemptNumber 클래스를 도메인에 만든 후, 검증 로직까지 이 클래스에서 진행하는 방식으로 구현했습니다!

validator 패키지는 사용하지 않게 되어 제거했습니다. 이 방식이 맞는 것인지 궁금합니다..!😭

skullkim commented on src/main/java/racingcar/util/generator/RandomNumberGenerator.java:
```diff
@@ -1,10 +1,17 @@
+
 public class RandomNumberGenerator {
-	private static final int MULTIPLY_NUMBER = 10;
+	private static final int RANDOM_BOUND = 10;
```
NameGenerator, NumberGenerator 클래스 이름에서 Generator라는 단어 대신 Convertor라는 단어를 상요하면 어떨까요? RandomNumberGenerator는 내부에서 난수를 생성해 반환하기 때문에 Generator가 어울리는 것 같아요. 하지만 NameGenerator와 NumberGenerator는 값을 받아서 변환해주는 역할을 하기 때문에 Convertor가 더 적절할거 같네요😄

pine14 replied:
값을 생성하는 게 아니라 Convertor 라는 이름이 더 적절할 것 같습니다! 
NameConvertor, NumberConvertor로 이름 변경하겠습니다👍

skullkim commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,52 @@
+		}
+	}
+
+	public int findMaxPosition() {
```
findMaxPosition() 메서드를 findWinnersName()에서만 사용하고 있음에도 public으로 열고 테스트를 진행했네요. 이렇게 테스트를 진행한 이유가 있을까요?

pine14 replied:
```java
@Test
@DisplayName("가장 멀리 간 위치 찾기")
void max_position() {
        for (int i = 0; i < 3; ++i) {
		car1.moveForward(new ForwardNumberGenerator());
	}
	car2.moveForward(new ForwardNumberGenerator());

	assertThat(cars.findMaxPosition()).isEqualTo(3);
}
```
라는 테스트에서 1번 자동차를 3번, 2번 자동차를 1번 전진시킨 후 가장 많이 전진한 위치가 3임을 검증하기 위한 테스트를 작성했었습니다!

skullkim replied:
findMaxPosition()이라는 메서드를 Cars 내부에서만 사용함에도 외부로 드러낸다면 테스트가 프로덕션 코드에 영향을 미치는 꼴이 되므로 옳지 않다는 생각이 드네요. 또 한, "위치로 우승자를 판별한다"는 내부에 감춰야 할 기능이고 "우승자의 이름을 찾는다"가 밖으로 드러나야 할 기능이라 이 부분을 왜 테스트 해야 하는지 모르겠네요🧐. 이 테스트를 작성한 이유에 대해 알려주실 수 있나요?

pine14 replied:
1단계 요구사항 중에 모든 로직에 단위 테스트를 구현한다는 글을 읽어서 해당 테스트를 작성했었습니다!

사실 이 메서드에 관한 고민이 있었는데 
```java
public List<String> findWinnersName() {
		return cars.stream()
			.filter(car -> car.isWinner(findMaxPosition()))
			.map(car -> car.getName().getValue())
			.collect(toList());
	}
```
이 메서드에서 isWinner 메서드를 사용하고 있었습니다.
```java
public boolean isWinner(int maxPosition) {
	return position.getValue() == maxPosition;
}
```
isWinner 메서드는 다음처럼 구현되어 있습니다. 

isWinner 메서드를 테스트할 때 maxPosition에 임의로 값을 지정해준 위치가 그 자동차의 위치와 같다면 isWinner메서드는 true를 반환할 것이고, 그렇다면 저 메서드를 잘못 설계한 것이 아닌가 하는 생각이 들었습니다.

findMaxPosition()을 findWinnersName()에서만 사용하게 하고(테스트를 삭제하고) private 메서드로 만든다면 말씀하신 것처럼 내부에 기능을 감출 수 있을 것 같습니다.

이런 식으로 한다면 findMaxPosition이라는 로직은 어떻게 테스트 할 수 있을까요?

https://fishbowl.pastiche.org/2003/03/28/testing_private_methods_dont_do_it

다음과 같이 private method들은 테스트하지 않는다는 글이 있었습니다.
```java
private int findMaxPosition() {
	return cars.stream()
			.mapToInt(car -> car.getPosition().getValue())
			.max()
			.orElse(0);
}
```
그렇다면 이 메서드를 믿고 쓸 수 있는지 궁금합니다.

findWinnersName() 메서드를 테스트 한 것으로 충분할까요..?

skullkim replied:
TDD를 하는 과정을 보면 요구사항 -> 테스트 코드 작성 순으로 진행되죠. 이 말을 객체지향적으로 생각한다면 사용자가 사용할 기능이 무엇인지를 파악하고 그 기능에 필요한 책임을 가질 객체들을 찾은 후, 그 책임들을 테스트로 작성하는게 되요. 그런대, 객체들은 본질적으로 메시지를 주고 받으면서 서로에게 각자가 가질 책임을 주죠. 그럴려면 이 메시지(메서드)는 public이어야 하고요. 그래서 public으로 된 부분만 체크해도 문제 없다고 생각해요.

위 코드의 경우 "우승자를 찾는다"가 객체의 책임이 되는 거고, 그 책임을 스스로 이행하는 도중에 최대 위치가 필요한거죠.

pine14 replied:
좋은 글이네요.. 너무 이해가 잘 됐습니다!!

skullkim reviewed (changes requested):
안녕하세요 의찬님~ 두 번째 단계도 잘 수행해 주셨네요😊. 질문주신 객체지향 생활체조 7번째 원칙에 대해 답변을 드리자면 저는 primitive type을 래핑한 클래스는 7 번째 원칙에서 말한 인스턴스 필드에 포함시키지 말아야 한다고 생각해요~.

객체지향 생활체조가 최초로 거론된 도서 "The ThoughtWorks Anthology"의 "Chapter 6.2 The Exercise"에서 "7. Don't use any classes with more than two instance variables"에 대한 이유를 다음과 같이 설명하고 있어요.
```
Adding a new instance variable to a
class immediately decreases the cohesion of that class.
```
즉, 인스턴스 필드를 많이 추가하면 응집도가 떨어진다는게 이유인대요. 하지만 primitive type을 래핑하는 이유는 원래 그 클래스에서 해당 primitive type이 필요하지만, 더 명확한 서술을 위해, 많아 봐야 검증 책임을 위해 한번 객체로 래핑을 하는대에 있죠. 실제로 위에서 말한 것과 같은책, 같은 챕터에서 객체지향 생활체조 3번째 원칙 "Wrap All Primitives and Strings"를 보면 다음과 같은 의견을 볼 수 있어요.
```
An int on its own is just a scalar with no meaning. When a method takes
an int as a parameter, the method name needs to do all the work of
expressing the intent. If the same method takes an hour as a parameter,
it’s much easier to see what’s happening. Small objects like this can
make programs more maintainable, since it isn’t possible to pass a year
to a method that takes an hour parameter.
```

Commit:
refactor : 변수명을 명사로 변경

클린 코드 2장에 따라 COMPILED_CUSTOM_SPLIT_REGEX, 즉 컴파일된 사용자 정의 분할 정규식으로 이름을
변경했습니다.

Commit:
refactor : static 예약어 제거

static

Commit:
refactor : 클래스 이름 변경

NameGenerator와 NumberGenerator는 값을 받아서 변환해주는 역할을 합니다.
따라서 Convertor라는 이름으로 변경했습니다.

Commit:
test : 랜덤 값 테스트를 위한 임시 moveForward1 메서드 구현

NumberGenerator를 인터페이스로 만들고 구현체인 ForwardNumberGenerator를 통해 자동차가 전진했을 때의 테스트를
진행했습니다.

Commit:
feat : RandomNumberGenerator를 NumberGenerator의 구현체로 변경

RandomNumberGenerator를 NumberGenerator의 구현체로 변경한 후 실제 프로덕션 코드에서는 랜덤으로 수를 생성하도록
하였습니다.

Commit:
test : StopNumberGenerator로 정지 테스트 구현

NumberGenerator의 구현체인 StopNumberGenerator로 자동차가 전진할 때 테스트를 작성했습니다.

Commit:
refactor : 매직 넘버 피하기

모든 넘버는 그 자체만으로 크기를 표기하는 것 외의 의미를 가지지 않습니다. 따라서 상수를 사용했습니다.

Commit:
refactor : convertor 패키지 생성 및 메서드명 변경

Generator 클래스를 Convertor 라는 이름으로 변경하였으므로 convertor라는 패키지를 만들어 따로 분리했습니다.
메서드 명도 convertor로 시작하게 변경했습니다.

Commit:
refactor : AttemptNumberValidator로 클래스명 변경

미션에는 자동차의 위치와 시도 횟수라는 Number가 있는데, NumberValidator라는 클래스명보다는
시도 횟수 검증에 관한 클래스이므로 AttemptNumberValidator라는 이름이 더 가독성과 유지보수에 좋다고 생각해서
변경했습니다.

Commit:
feat : 시도 횟수를 입력받는 로직 도메인 패키지로 이동

시도 횟수를 정하는 것은 도메인 로직이므로 도메인으로 옮기는 방식으로 변경했습니다.

Commit:
feat : AttemptNumber 클래스 생성

시도 횟수 객체에서 검증까지 진행 한 후 컨트롤러에서 사용하는 방식으로 구현했습니다.

skullkim commented on src/main/java/racingcar/controller/RaceGameController.java:
```diff
@@ -1,63 +1,67 @@
-		} catch (IllegalArgumentException e) {
-			OutputData.printErrorMessage(e.getMessage()); // [ERROR] + 메시지 출력
-			addCarNamesToCars();
+			OutputView.printCarsPosition(cars.getCars());
```
"게임 진행 가능 횟수 만큼 게임을 진행한다"는 도메인 이라는 생각이 드는대요. 그래서 컨트롤러는 "게임 진행 가능 횟수"에 대한 변화를 몰라야 한다고 생각해요. 의찬님은 어떻게 생각하시나요?

pine14 replied:
저도 도메인 로직이라고 생각합니다.

AttemptNumber 클래스를 Cars 클래스의 필드로 둘까 생각했는데, 그렇게 구현하면 일급 컬렉션의 정의에서 벗어나므로 시도하지 않았습니다.

Cars 일급 컬렉션과 AttemptNumber를 필드로 가지는 Racing 클래스를 만들어 자동차를 전진시켰는데, 문제는 이 도메인에서 뷰를 사용하고 있다는 점입니다..

컨트롤러가 게임 진행 가능 횟수에 관한 내용은 모르게 구현한 것 같은데, 제가 생각하기로는 게임 한 번 진행하고 그 위치를 출력하고 또 게임 한 번 진행하고 위치를 출력하고 이런식으로 이뤄지기 때문에 위치를 출력하는 View를 어떻게 분리해야할지 감이 잘 잡히지 않습니다. 

어떻게 하면 view 부분을 잘 분리할 수 있을까요..?😭

skullkim commented on src/main/java/racingcar/util/generator/ForwardNumberGenerator.java:
```diff
@@ -0,0 +1,10 @@
+package racingcar.util.generator;
+
+public class ForwardNumberGenerator implements NumberGenerator {
```
이 클래스는 테스트에서만 사용하고 있는대 프로덕션 패키지에 넣는게 맞을까요?

pine14 replied:
테스트 패키지로 이동하는 것이 맞는 것 같습니다!

skullkim commented on src/main/java/racingcar/util/generator/StopNumberGenerator.java:
```diff
@@ -0,0 +1,10 @@
+package racingcar.util.generator;
+
+public class StopNumberGenerator implements NumberGenerator {
```
이 클래스도 테스트에서만 사용하고 있네요.

pine14 replied:
테스트 패키지로 이동시키겠습니다!

skullkim reviewed (changes requested):
의찬님~ 리뷰 잘 반영해 주셨네요. 추가 리뷰 남겼으니 확인해주세요~

Commit:
refactor : findMaxPosition() 메서드의 접근 제어자 변경

위치로 우승자를 판별한다는 내부에 감춰야 할 기능이므로 접근 제어자를 private으로 변경했습니다.

Commit:
refactor : 테스트에서만 사용하는 클래스의 패키지 변경

ForwardNumberGenerator, StopNumberGenerator 클래스는 테스트에서만 사용하므로
테스트 패키지로 이동시켰습니다.

Commit:
refactor : getter 체이닝 변경

Car 클래스에서 getPosition()을 this.position.getValue()를 리턴하도록 해 getter 체이닝을 제거했습니다.

Commit:
refactor : AttemptNumber 클래스의 책임 분리

AttemptNumber 클래스가 String 타입에서 int 타입으로 변환하는 책임까지 가지고 있으므로
AttemptNumberConvertor라는 클래스를 만들어 책임을 분리했습니다.

Commit:
refactor : Racing 클래스 구현

컨트롤러에서 게임 진행 횟수라는 도메인을 알고 있기 때문에 Racing 클래스를 만들어 분리했습니다.

Commit:
refactor : getter 체이닝 변경

Car 클래스에서 getName()을 this.name.getValue()를 리턴하도록 해 getter 체이닝을 제거했습니다.
